### DISCUSSION TITLE
How to update from Aspell 0.60.7 to 0.60.8.1?

### DISCUSSION THREAD

**Discussion by elchenberg** [renovatebot/renovate#32216] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

I created a custom datasource for Aspell nearly identical to the [Nginx example from the documentation](https://docs.renovatebot.com/modules/datasource/custom/#nginx-directory-listing):
```json
{
  "packageRules": [
    "matchDatasources": ["custom.aspell"],
    "extractVersion": "^aspell-(?<version>.+)\\.tar\\.gz$",
  ],
  "customDatasources": {
    "aspell": {
      "defaultRegistryUrlTemplate": "https://ftp.gnu.org/gnu/aspell",
      "format": "html",
    },
  }
}
```

This seems to work very well. I think my problem is Aspell's versioning.

I have Aspell 0.60.7. Aspell 0.60.8 was released on 2019-10-12 and 0.60.8.1 was released on 2023-12-19. With the configuration above Renovate would update me to 0.60.8 but not to 0.60.8.1.

I tried to add regex versioning (`"versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)(?:\\.(?<patch>\\d+)(?:\\.(?:<build>\\d+))?)?$"
`) to the package rule to consider the `x.x.x.1` as a `build` version. But this does not work and I think the reason is that the previous version did not have a build number.

So I would like to update from 0.60.8 to 0.60.8.1 and if there is a new patch release 0.60.9 in the future I would like to update to it, too, even if it does not have a build number.

Is there a way to do this with Renovate?

### Logs (if relevant)

<details><summary>Logs</summary>

```
DEBUG: packageFiles with updates (repository=local)
       "config": {
         "regex": [
           {
             "deps": [
               {
                 "depName": "aspell",
                 "currentValue": "0.60.7",
                 "datasource": "custom.aspell",
                 "replaceString": "# renovate: datasource=custom.aspell depName=aspell\nENV ASPELL_VERSION=0.60.7\n",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "0.60.8",
                     "newValue": "0.60.8",
                     "newMajor": 0,
                     "newMinor": 60,
                     "newPatch": 8,
                     "updateType": "patch",
                     "branchName": "renovate/aspell-0.x"
                   }
                 ],
                 "packageName": "aspell",
                 "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)(?:\\.(?<patch>\\d+)(?:\\.(?:<build>\\d+))?)?$",
                 "warnings": [],
                 "currentVersion": "0.60.7",
                 "isSingleVersion": true,
                 "fixedVersion": "0.60.7"
               }
             ],
           }
         ]
       }
```

</details>

**Comment by rarkins** [comment, marked as answer]
Use regex101.com to test your versioning regex (you will need to replace double `\\` with `\` though). You have a typo in the build part

**Reply by elchenberg** [reply]
Thank you!
